depinject/container_test.go

```diff
@@ -644,35 +644,124 @@ type AlsoDuck interface {
 }
 
 type Mallard struct{}
+type Canvasback struct{}
 
-func (duck Mallard) quack() {}
-
-type KeyedOutput struct {
-	depinject.Out
-	Duck Duck `key:"foo"`
-}
-
-type KeyedInput struct {
-	depinject.In
-	AlsoDuck AlsoDuck `key:"foo"`
-}
+func (duck Mallard) quack()    {}
+func (duck Canvasback) quack() {}
 
 type Pond struct {
 	Duck AlsoDuck
 }
 
-func TestKeyedInputOutput(t *testing.T) {
+func TestImplicitAndExplicitBindings(t *testing.T) {
 	var pond Pond
 
 	require.NoError(t,
 		depinject.Inject(
 			depinject.Provide(
-				func() KeyedOutput { return KeyedOutput{Duck: Mallard{}} },
-				func(in KeyedInput) Pond {
-					require.NotNil(t, in.AlsoDuck)
-					return Pond{Duck: in.AlsoDuck}
+				func() Mallard { return Mallard{} },
+				func(duck Duck) Pond {
+					require.NotNil(t, duck)
+					return Pond{Duck: duck}
 				}),
 			&pond))
 
 	require.NotNil(t, pond)
+
+	multImplProvider := depinject.Provide(
+		func() Mallard { return Mallard{} },
+		func() Canvasback { return Canvasback{} },
+		func(duck Duck) Pond {
+			return Pond{Duck: duck}
+		})
+	canvasbackPref := depinject.Prefer(
+		"github.com/cosmos/cosmos-sdk/depinject_test/depinject_test.Duck",
+		"github.com/cosmos/cosmos-sdk/depinject_test/depinject_test.Canvasback")
+
+	require.ErrorContains(t, depinject.Inject(multImplProvider, &pond), "Multiple implementations")
+
+	require.NoError(t,
+		depinject.Inject(
+			depinject.Configs(canvasbackPref, multImplProvider),
+			&pond))
+
+	require.IsType(t, pond.Duck, Canvasback{})
+
+}
+
+func TestExplicitModuleBindings(t *testing.T) {
+	var pond Pond
+
+	require.ErrorContains(t,
+		depinject.Inject(
+			depinject.Configs(
+				depinject.PreferInModule(
+					"foo",
+					"github.com/cosmos/cosmos-sdk/depinject_test/depinject_test.Duck",
+					"github.com/cosmos/cosmos-sdk/depinject_test/depinject_test.DoesNotExist"),
+				depinject.Provide(
+					func(depinject.ModuleKey) Mallard { return Mallard{} },
+					func(depinject.ModuleKey) Canvasback { return Canvasback{} }),
```

Comment:
```suggestion
					func() Mallard { return Mallard{} },
					func() Canvasback { return Canvasback{} }),
```
no reason we need to make these module-scoped deps